Slot Swap Option

zoracon opened the pull request:
Feel free to reject if not wanted. I was playing with the option to switch PIV slots with generate function 

I built and tested locally on my Ubuntu (20.04 ) based laptop on snap installed Go (1.18.5)

Great tool!

I am also do not regularly program in Go. So forgive any glaring missteps. Feedback welcome if you have the time.

yoavamit commented:
That seems like a nice feature!

For context - when we first started this project we found out that different slots have [different restrictions and use cases](https://developers.yubico.com/PIV/Introduction/Certificate_slots.html).
For example, 9e, the "Card Authentication" slot is the only slot that doesn't require a PIN to access the private key when signing with it.
We chose that slot to reduce friction for the end user.

I think that adding the option to use different slots is great here, especially if it helps other people and organizations adopt this tool.

yoavamit commented on cmd/pivit/main.go:
```diff
@@ -21,6 +21,7 @@ func runCommand() error {
 	verifyFlag := getopt.BoolLong("verify", 0, "verify a signature")
 	resetFlag := getopt.BoolLong("reset", 'r', "resets the smart card PIV applet and sets new PIN, random PUK, and PIN derived management key")
 	generateFlag := getopt.BoolLong("generate", 'g', "generates a new key pair and a certificate signing request")
+	slotSwapGenerate := getopt.StringLong("slot-swap", 'w', "", "choose 4 available PIV slots, defaults to PIV slot 9e", "slot")
```
I think "slot" would be a better name, but I'm open to suggestions :)
You can also set the default value here to `"9e"` on the third parameter

yoavamit commented on cmd/pivit/generate.go:
```diff
@@ -19,7 +19,7 @@ import (
 
 // commandGenerate generates a new keypair and certificate signing request
-func commandGenerate() error {
+func commandGenerate(swap string) error {
```
If we're **generating** a key pair in a specific slot, we probably want to later be able to specify that slot when performing other commands like `import`, `print` and `sign`

yoavamit commented on cmd/pivit/generate.go:
```diff
@@ -41,13 +41,20 @@ func commandGenerate() error {
 		return errors.Wrap(err, "get pin")
 	}
 
+	slot := piv.SlotCardAuthentication
```
nit: a slighly nicer way to write this would be to use something like:
```
type SlotID string

var slotMap = map[SlotID]piv.Slot{
    piv.SlotCardAuthentication.String(): piv.SlotCardAuthentication,
    piv.SlotAuthentication.String():     piv.SlotAuthentication,
    piv.SlotSignature.String():          piv.SlotSignature,
    piv.SlotKeyManagement.String():      piv.SlotKeyManagement, 
}

zoracon commented:
> That seems like a nice feature!
> 
> For context - when we first started this project we found out that different slots have [different restrictions and use cases](https://developers.yubico.com/PIV/Introduction/Certificate_slots.html). For example, 9e, the "Card Authentication" slot is the only slot that doesn't require a PIN to access the private key when signing with it. We chose that slot to reduce friction for the end user.
> 
> I think that adding the option to use different slots is great here, especially if it helps other people and organizations adopt this tool.

@yoavamit Sounds great. Looks like I should sprinkle in some more documentation if the user plans to use different slots. Personally, I normally utilize slot 9a and 9c. I would totally use this for attestation on different slots. I will be able to actually retest that hypothesis out after November 15th with a CA issued code signing cert(New CA/B forum requirements kicking in for CAs and code signing certs). It would make my life a lot easier in the provisioning realm.

p.s. sorry for the choppy git history, I meant to clean that up before and forgot when I made a PR. Requesting a squash merge if it gets approved for merge.

zoracon commented:
@yoavamit So I addressed most comments. However, I ran into an issue with modules and methods. I have slot management in `import`, `print`, and `sign`. However, I violate DRY principle here a bit since I was not able to use my `utils` methods without build errors. 

I assume this may be due to my lack of Go knowledge. I read over how to imports and packages but feel like I am missing something simple here.

zoracon commented:
As an update. I stepped away and came back to my local `go.mod` to see how to properly map my own utils to locally test before pushing up and found out what my issues were. Pushed up with fixes removing all duplicate code.

yoavamit commented:
> As an update. I stepped away and came back to my local `go.mod` to see how to properly map my own utils to locally test before pushing up and found out what my issues were. Pushed up with fixes removing all duplicate code.

👀 I'll review it soon.
In the meantime, could you please also sign the CLA mentioned [here](https://github.com/cashapp/pivit/blob/main/CONTRIBUTING.md).

yoavamit commented on cmd/pivit/sign.go:
```diff
@@ -12,18 +12,17 @@ import (
 
 // commandSign signs the filename given in fileArgs or the content from stdin if no filename was supplied
-func commandSign(statusFd int, detach, armor bool, userId, timestampAuthority string, fileArgs []string) error {
+func commandSign(statusFd int, detach, armor bool, userId, timestampAuthority string, slot string, fileArgs []string) error {
```
Notice that when signing, we use `yubikey.Sign` in the `yubikey_identity.go` file.
That function specifies `piv.KeyAuth{PINPolicy: piv.PINPolicyNever}` when accessing the private key in the PIV applet.
It also hardcodes `piv.SlotCardAuthentication` as the slot to use.

If we're specifying a different slot to use when signing, we'll need to make some changes there as well.

I suggest doing the following changes in the `yubikey/yubikey_identity.go` file:
1. Add a slot variable to the `YubikeySigner` type.
2. Change `YubiukeySigner.Sign` and `yubikeySigner.Public` accordingly so that they'll use the right slot
3. Add some logic to `YubiukeySigner.Sign` so that if using any slot other than `9e`, it'll prompt for a PIN.

zoracon replied:
That makes sense. I am fishtailing into the holidays and run the reigns over the 2 open source projects, so I might not be able to take a look until later next week. I want to see this feature in time to use this tool for wide scale provisioning, so it's on my list of importance.

yoavamit reviewed (changes requested):
I feel like this is almost done, but we *must* change the `YubikeySigner` struct as well.
Otherwise, it'll break the signing functionality when using other slots.

yoavamit commented on cmd/pivit/yubikey/yubikey_identity.go:
```diff
@@ -16,37 +17,52 @@ func Yubikey() (*piv.YubiKey, error) {
+		return nil, errors.Wrap(err, "get pin")
+	}
+
+	auth := piv.KeyAuth{
+		PINPolicy: piv.PINPolicyAlways,
+		PINPrompt: func() (string, error) {
+			fmt.Println("Touch Yubikey now to sign data...")
+			return pin, nil
+		},
+	}
```
Instead of prompting for the PIN at the top of this function, it should be done here
```suggestion
	auth := piv.KeyAuth{
		PINPolicy: piv.PINPolicyAlways,
		PINPrompt: func() (string, error) {
		        pin, err := utils.GetPin()
			if err != nil {
				return "", errors.Wrap(err, "get pin")
			}
			fmt.Println("Touch Yubikey now to sign data...")
			return pin, nil
		},
	}
```

yoavamit commented on cmd/pivit/yubikey/yubikey_identity.go:
```diff
@@ -16,37 +17,52 @@ func Yubikey() (*piv.YubiKey, error) {
 func (y YubikeySigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
-	private, err := y.yk.PrivateKey(piv.SlotCardAuthentication, y.Public(), piv.KeyAuth{PINPolicy: piv.PINPolicyNever})
+
+	// Caching for PINPolicyOnce isn't supported on YubiKey versions older than 4.3.0.
+	// If specified, a PIN will be required for every operation.
+	// see: https://github.com/go-piv/piv-go/blob/v1.10.0/piv/key.go#L446
+	pin, err := utils.GetPin()
+	if err != nil {
+		return nil, errors.Wrap(err, "get pin")
+	}
```
```suggestion
```

zoracon commented on cmd/pivit/yubikey/yubikey_identity.go:
```diff
@@ -16,37 +17,52 @@ func Yubikey() (*piv.YubiKey, error) {
 func (y YubikeySigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
-	private, err := y.yk.PrivateKey(piv.SlotCardAuthentication, y.Public(), piv.KeyAuth{PINPolicy: piv.PINPolicyNever})
+
+	// Caching for PINPolicyOnce isn't supported on YubiKey versions older than 4.3.0.
```
I am just reflecting some of the Yubico documentation here. I can remove since there wasn't a particular reason other than that.

zoracon commented:
Ah I was going to address the nits but saw you merged. Thanks for feedback along the way! And I plan to learn a little more Go after this. This was a good exercise for me and can't wait to use this tool the near future for provisioning Yubikeys.
